sage_code/strong_uniform_bounds.py

```diff
@@ -109,20 +123,71 @@ def bound_from_split_type(split_type, eps, q):
         if pil_int == 0:
             zero_detection_flag = True
         else:
-            running_lcm = lcm(pil_int, running_lcm)
+            pil_int = gcd(known_mult_bound, pil_int)
+            running_lcm = gcd(known_mult_bound, lcm(pil_int, running_lcm))
     if zero_detection_flag:
         return running_lcm, 0
     else:
         return running_lcm, running_lcm
 
 
-def B_eps_q(d,eps,q):
+def B_eps_q(d,eps,q, known_mult_bound=0):
 
     split_types = splitting_types(d)
     B_star = 1
     B = 1
     for split_type in split_types:
-        pil_int_star, pil_int = bound_from_split_type(split_type, eps, q)
-        B_star = lcm(B_star,pil_int_star)
-        B = lcm(B,pil_int)
-    return B_star, B
\ No newline at end of file
+        pil_int_star, pil_int = bound_from_split_type(split_type, eps, q, known_mult_bound)
+        B_star = gcd(known_mult_bound, lcm(B_star,pil_int_star))
+        B = gcd(known_mult_bound, lcm(B,pil_int))
+    return B_star, B
+
+def core_loop(d, eps, aux_primes):
+
+    mult_upper_bd = 0
+
+    for q in aux_primes:
+        B_star, B = B_eps_q(d,eps,q,mult_upper_bd)
+        assert B_star == B
+        mult_upper_bd = gcd(mult_upper_bd, B_star)
+
+    return mult_upper_bd
+
+
+def unif_bd(d, eps):
+
+    tr_eps = sum(eps)
+
+    if tr_eps % 6 != 0:
+        aux_primes = prime_range(6)
+    elif tr_eps % 12 == 6:
+        aux_primes = prime_range(5, 15)
+    else:
+        raise ValueError("can't deal with this case")
+
+    return core_loop(d, eps, aux_primes)
+
+
+def type_one_unif_primes(d, q_bd=4):
+
+    assert q_bd > 3
+
+    aux_primes = prime_range(3, q_bd+1)
+
+    mult_upper_bd = 0
+    type_one_eps =  (0,) * d
+    bad_formal_immersion_list, bad_aux_prime_dict = cached_bad_formal_immersion_data(d)
+
+    for q in aux_primes:
+        B_star, B = B_eps_q(d,type_one_eps,q,mult_upper_bd)
+
+        agfi_q = bad_aux_prime_dict.get(str(q),1)
+
+        q_prod = prod([(q ** f - 1) for f in range(1,d+1)])
```

Comment:
I think `q_prod` isn't actually necessary. In fact we only care about the prime divisors so we could do instead.
```suggestion
        q_lcm = lmc([(q ** f - 1) for f in range(1,d+1)])
```

However q_lcm will be a divisor of B_star. So this entire line can be removed.

Reply:
> However q_lcm will be a divisor of B_star. So this entire line can be removed.

That might be true, but I'm don't quite see it; essentially because of the 12e_i exponent in the B_{eps,q,\beta} integer in step 2 of Algorithm 5.1. If it can be removed, then the statement of Proposition 5.3 can be cleaned up. Let's fix it there first, before fixing it here ...